1) Download the customer data set and add it to your “Original Data” folder.

2) Create a new notebook in your “Scripts” folder for part 1 of this task.

3) Import your analysis libraries, as well as your new customer data set as a dataframe.

4) Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

5) Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

6) Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the same data type!


# -----Task----- 

# 3. Import Libraries and Customer Dataset

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Creating a path to to .csv file
path = r'C:\Users\mmoss\20-12-2021 Instacart Basket Analysis'

In [3]:
# Importing the .csv file
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
# Importing the prepared Instacart data
df_prepared = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', '4.8_ords_products_merged_derived_grouped.pkl'))

# 4. Wrangling the data

In [5]:
# Displaying the data to see the columns
df_customers.head(20)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514

Columns to rename: First_Name to first_name, 
                   Surnam to surname, 
                   Gender to gender, 
                   STATE to state, 
                   Age to age, 
                   n_dependants to number_of_dependants, 
                   fam_status to marital_status, 

In [6]:
# First Name
df_customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [7]:
# Surnam
df_customers.rename(columns = {'Surnam' : 'surname'}, inplace = True)

In [8]:
# Gender
df_customers.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [9]:
# STATE
df_customers.rename(columns = {'STATE' : 'state'}, inplace = True)

In [10]:
# Age
df_customers.rename(columns = {'Age' : 'age'}, inplace = True)

In [11]:
# n_dependants
df_customers.rename(columns = {'n_dependants' : 'number_of_dependants'}, inplace = True)

In [12]:
# fam_status
df_customers.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [13]:
# Test it out
df_customers.head(20)

user_id first_name   surname  gender           state  age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    number_of_dependants                    marital_status  income  
0                      3                           married  165665  
1                      0                            single   59285  
2                      2                           married   99568  
3                      0                            single   42049  
4                      1                           married   40374  
5                      2                           married   49643  
6                      0                            single   61746  
7                      0                            single   63712  
8                      1                           married  162432  
9                      2                           married   32072  
10                     2  living with parents and siblings   43443  
11                     2                           married   47988  
12                     1                           married   57835  
13                     1                           married   71190  
14                     2                           married  146559  
15                     0                  divorced/widowed  100091  
16                     3                           married   57096  
17                     1                           married   91826  
18                     2                           married   91111  
19                     1                           married   34514

Success!

All columns here look useful. Will not remove any. 

# 5. Data quality and consistency checks

In [14]:
# Checking for missing values
df_customers.isnull().sum()

user_id                     0
first_name              11259
surname                     0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
marital_status              0
income                      0
dtype: int64

In [15]:
#Create a subset of just the missing values
df_nan = df_customers[df_customers['first_name'].isnull() == True]

In [16]:
#display it
df_nan

user_id first_name   surname  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependants marital_status  income  
53                         2        married   41709  
73                         0         single   82518  
82                         3        married  117099  
99                         1        married  155673  
105                        3        married  151819  
...                      ...            ...     ...  
206038                     3        married   87783  
206044                     2        married  108488  
206090                     1        married   45275  
206162                     0         single   56325  
206171                     2        married   59222  

[11259 rows x 10 columns]

first_name has 11,259 rows with missing names. It would be easier to filter out these values and make a new dataframe just with the records that have first names.

In [17]:
# Running shape to see a before and after row count
df_customers.shape

(206209, 10)

In [18]:
# Removing the missing value rows
df_customers_clean = df_customers[df_customers['first_name'].isnull() == False]

In [19]:
# Testing the shape
df_customers_clean.shape

(194950, 10)

Success! Missing rows have been removed.

In [20]:
# Checking for duplicates
df_dups = df_customers_clean[df_customers_clean.duplicated()]

In [21]:
# Displaying them
df_dups

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, number_of_dependants, marital_status, income]
Index: []

There are no duplicates so I will not need to address any.

In [22]:
# Checking for mixed type data
df_customers_clean.head(20)

user_id first_name   surname  gender           state  age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    number_of_dependants                    marital_status  income  
0                      3                           married  165665  
1                      0                            single   59285  
2                      2                           married   99568  
3                      0                            single   42049  
4                      1                           married   40374  
5                      2                           married   49643  
6                      0                            single   61746  
7                      0                            single   63712  
8                      1                           married  162432  
9                      2                           married   32072  
10                     2  living with parents and siblings   43443  
11                     2                           married   47988  
12                     1                           married   57835  
13                     1                           married   71190  
14                     2                           married  146559  
15                     0                  divorced/widowed  100091  
16                     3                           married   57096  
17                     1                           married   91826  
18                     2                           married   91111  
19                     1                           married   34514

All columns have just one type of data in them. No need to address mixed-type columns.

# 6. Combining your data

In [23]:
# Displaying df_customers
df_customers.head(10)

user_id first_name   surname  gender           state  age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   number_of_dependants marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374  
5                     2        married   49643  
6                     0         single   61746  
7                     0         single   63712  
8                     1        married  162432  
9                     2        married   32072

In [24]:
#Displaying df_prepared
df_prepared.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   
5                    19.0        False         196                  1   
6                    20.0        False         196                  1   
7                    14.0        False         196                  2   
8                     0.0        False         196                  4   
9                    30.0        False         196                  1   

   reordered  ... product_name aisle_id  department_id  prices  \
0          0  ...         Soda       77              7     9.0   
1          1  ...         Soda       77              7     9.0   
2          1  ...         Soda       77              7     9.0   
3          1  ...         Soda       77              7     9.0   
4          1  ...         Soda       77              7     9.0   
5          1  ...         Soda       77              7     9.0   
6          1  ...         Soda       77              7     9.0   
7          1  ...         Soda       77              7     9.0   
8          1  ...         Soda       77              7     9.0   
9          1  ...         Soda       77              7     9.0   

     price_range_loc     busiest_day   busiest_day_2 busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy         Fewest orders   
1  Mid-range product  Regularly busy    Slowest days         Fewest orders   
2  Mid-range product  Regularly busy    Slowest days         Fewest orders   
3  Mid-range product      Least busy  Regularly busy         Fewest orders   
4  Mid-range product      Least busy  Regularly busy         Fewest orders   
5  Mid-range product  Regularly busy  Regularly busy         Fewest orders   
6  Mid-range product  Regularly busy  Regularly busy         Fewest orders   
7  Mid-range product  Regularly busy  Regularly busy         Fewest orders   
8  Mid-range product  Regularly busy  Regularly busy         Fewest orders   
9  Mid-range product      Least busy  Regularly busy         Fewest orders   

  average_price  spending_flag  
0      6.367797    Low Spender  
1      6.367797    Low Spender  
2      6.367797    Low Spender  
3      6.367797    Low Spender  
4      6.367797    Low Spender  
5      6.367797    Low Spender  
6      6.367797    Low Spender  
7      6.367797    Low Spender  
8      6.367797    Low Spender  
9      6.367797    Low Spender  

[10 rows x 21 columns]

In [25]:
df_merged = df_prepared.merge(df_customers, on = 'user_id')

In [26]:
df_merged

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32399727    156685   106143            26                   4   
32399728    484769    66343             1                   6   
32399729   1561557    66343             2                   1   
32399730    276317    66343             3                   6   
32399731   2922475    66343             4                   1   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True         196   
1                         7                    15.0        False         196   
2                        12                    21.0        False         196   
3                         7                    29.0        False         196   
4                        15                    28.0        False         196   
...                     ...                     ...          ...         ...   
32399727                 23                     5.0        False       19675   
32399728                 11                     NaN         True       47210   
32399729                 11                    30.0        False       47210   
32399730                 15                    19.0        False       47210   
32399731                 12                    30.0        False       47210   

          add_to_cart_order  reordered  ... spending_flag  first_name  \
0                         1          0  ...   Low Spender       Linda   
1                         1          1  ...   Low Spender       Linda   
2                         1          1  ...   Low Spender       Linda   
3                         1          1  ...   Low Spender       Linda   
4                         1          1  ...   Low Spender       Linda   
...                     ...        ...  ...           ...         ...   
32399727                  1          1  ...  High Spender      Gerald   
32399728                  1          0  ...   Low Spender  Jacqueline   
32399729                  1          1  ...   Low Spender  Jacqueline   
32399730                  1          1  ...   Low Spender  Jacqueline   
32399731                  1          1  ...   Low Spender  Jacqueline   

          surname  gender      state age date_joined number_of_dependants  \
0          Nguyen  Female    Alabama  31   2/17/2019                    3   
1          Nguyen  Female    Alabama  31   2/17/2019                    3   
2          Nguyen  Female    Alabama  31   2/17/2019                    3   
3          Nguyen  Female    Alabama  31   2/17/2019                    3   
4          Nguyen  Female    Alabama  31   2/17/2019                    3   
...           ...     ...        ...  ..         ...                  ...   
32399727    Yates    Male     Hawaii  25   5/26/2017                    0   
32399728   Arroyo  Female  Tennessee  22   9/12/2017                    3   
32399729   Arroyo  Female  Tennessee  22   9/12/2017                    3   
32399730   Arroyo  Female  Tennessee  22   9/12/2017                    3   
32399731   Arroyo  Female  Tennessee  22   9/12/2017                    3   

         marital_status  income  
0               married   40423  
1               married   40423  
2               married   40423  
3               married   40423  
4               married   40423  
...                 ...     ...  
32399727         single   53755  
32399728        married   46151  
32399729        married   46151  
32399730        married   46151  
32399731        married   46151  

[32399732 rows x 30 columns]

In [27]:
df_merged['_merge'].value_counts()

both          32399732
left_only            0
right_only           0
Name: _merge, dtype: int64

Success!

#### Exporting to .pkl file

In [28]:
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', '4.9_data_merged.pkl'))